README.md
# check_apache_status.pl

Plugin for Icinga, Nagios and Shinken to check the Apache httpd status.

If the Apache httpd status page is accessible then this plugin checks for following metrics:

- open slots
- busy workers
- idle workers
- requests per second
- bytes per second
- bytes per request



## Command Line Options

Usage: check_apache_status.pl [OPTIONS]

 -?, --usage

   Print usage information

 -h, --help

   Print detailed help screen

 -V, --version

   Print version information

 --extra-opts=[section][@file]

   Read options from an ini file. See https://www.monitoring-plugins.org/doc/extra-opts.html
   for usage and examples.

 -H, --hostname=STRING

   hostname or ip address to check

 -p, --port=INTEGER

   port, default 80 (http) or 443 (https)

 -U, --username

  username for basic auth, requires password

 -P, password

  password for basic auth, requires username

 -u, --uri=STRING

   uri, default /server-status

 -s, --ssl

   use https instead of http

 -N, --no\_validate

   do not validate the SSL certificate chain

 -R, --unreachable

   CRITICAL if socket timed out or http code >= 500

 -w, --warning=STRING

   warning threshold

 -c, --critical=STRING

   critical threshold

 -t, --timeout=INTEGER

   Seconds before plugin times out (default: 15)

 -v, --verbose

   Show details for command-line debugging (can repeat up to 3 times)



## Threshold and ranges

A threshold is a [range](https://www.monitoring-plugins.org/doc/guidelines.html#THRESHOLDFORMAT) with an alert level (either WARNING or CRITICAL).
Thresholds will be set comma separated in the following order:

- open slots
- busy workers
- idle workers
- requests per second
- bytes per second
- bytes per request

A range is defined as a start and end point (inclusive) on a numeric scale (possibly negative or positive infinity).

This is the generalised format for ranges:

```[@]start:end```

### Notes

- start ≤ end
- start and ":" is not required if start=0
- if range is of format "start:" and end is not specified, assume end is infinity
- to specify negative infinity, use "~"
- alert is raised if metric is outside start and end range (inclusive of endpoints)
- if range starts with "@", then alert if inside this range (inclusive of endpoints)

### Examples


Setting non for one of them is also possible. The following example sets the threshold for _busy workers_ to WARNING for at least 10 _busy workers_ and CRITICAL to at least of 25 _busy workers_:

```check_apache_status.pl -H localhost -w ,10,,,, -c ,25,,,,```

To additionally set the threshold for _open_slots_ to WARNING when at or below 10 and CRITICAL when at or below 5:

```check_apache_status.pl -H localhost -w 10:,10,,,, -c 5:,25,,,,```


CHANGELOG.md
# Change Log

## [1.4.3](https://github.com/lbetz/check_apache_status/tree/1.4.3) (2023-04-14)
[Full Changelog](https://github.com/lbetz/check_apache_status/compare/v1.4.2...1.4.3)

**Implemented enhancements:**

- Rename RPM package name [\#15](https://github.com/lbetz/check_apache_status/pull/15) ([lbetz](https://github.com/lbetz))
- Feature/debpkg [\#14](https://github.com/lbetz/check_apache_status/pull/14) ([lbetz](https://github.com/lbetz))

**Fixed bugs:**

- Extend status?auto to get the whole Scoreboard [\#13](https://github.com/lbetz/check_apache_status/pull/13) ([mdicss](https://github.com/mdicss))

## [v1.4.2](https://github.com/lbetz/check_apache_status/tree/v1.4.2) (2020-04-20)
[Full Changelog](https://github.com/lbetz/check_apache_status/compare/1.4.1...v1.4.2)

**Fixed bugs:**

- fix for SSLv3 errors, from https://stackoverflow.com/a/51128762 [\#12](https://github.com/lbetz/check_apache_status/pull/12) ([willhaines](https://github.com/willhaines))
- Fix no\_validate [\#11](https://github.com/lbetz/check_apache_status/pull/11) ([TheLastProject](https://github.com/TheLastProject))

## [1.4.1](https://github.com/lbetz/check_apache_status/tree/1.4.1) (2017-11-19)
[Full Changelog](https://github.com/lbetz/check_apache_status/compare/1.4.0...1.4.1)

**Fixed bugs:**

- Detect Idle Waits [\#9](https://github.com/lbetz/check_apache_status/issues/9)
- fix \#9 detect idle waits [\#10](https://github.com/lbetz/check_apache_status/pull/10) ([lbetz](https://github.com/lbetz))

## [1.4.0](https://github.com/lbetz/check_apache_status/tree/1.4.0) (2017-10-14)
[Full Changelog](https://github.com/lbetz/check_apache_status/compare/1.3.0...1.4.0)

**Implemented enhancements:**

- Exit plugin with CRITICAL if no response [\#3](https://github.com/lbetz/check_apache_status/issues/3)

**Closed issues:**

- Open slots threshold not working correctly [\#4](https://github.com/lbetz/check_apache_status/issues/4)
- Thresholds do not work properly [\#7](https://github.com/lbetz/check_apache_status/issues/7)
- "Use of uninitialized value in sprintf" in Performance.pm [\#5](https://github.com/lbetz/check_apache_status/issues/5)

**Merged pull requests:**

- Feature/add more metrics [\#8](https://github.com/lbetz/check_apache_status/pull/8) ([ghost](https://github.com/ghost))
- Proposed fix for \#5 [\#6](https://github.com/lbetz/check_apache_status/pull/6) ([ghost](https://github.com/ghost))

## [1.3.0](https://github.com/lbetz/check_apache_status/tree/1.3.0) (2017-06-06)
[Full Changelog](https://github.com/lbetz/check_apache_status/compare/1.2.0...1.3.0)

## [1.2.0](https://github.com/lbetz/check_apache_status/tree/1.2.0) (2017-02-17)
**Closed issues:**

- Hostname verification when using HTTPS [\#2](https://github.com/lbetz/check_apache_status/issues/2)
-  use Monitoring::Plugin;    SyntaxError: invalid syntax [\#1](https://github.com/lbetz/check_apache_status/issues/1)



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


check_apache_status.pl
#!/usr/bin/perl


use Monitoring::Plugin;
use Monitoring::Plugin::Getopt;
use Monitoring::Plugin::Threshold;
use LWP::UserAgent;
use HTTP::Status qw(:constants :is status_message);

our $VERSION = '1.4.3';

our ( $plugin, $option );

$plugin = Monitoring::Plugin->new( shortname => '' );


$options = Monitoring::Plugin::Getopt->new(
  usage   => 'Usage: %s [OPTIONS]',
  version => $VERSION,
  url     => 'https://github.com/lbetz/check_apache_status',
  blurb   => 'Check apache server status',
);

$options->arg(
  spec     => 'hostname|H=s',
  help     => 'hostname or ip address to check',
  required => 1,
);

$options->arg(
  spec     => 'port|p=i',
  help     => 'port, default 80 (http) or 443 (https)',
  required => 0,
);

$options->arg(
  spec     => 'uri|u=s',
  help     => 'uri, default /server-status?auto',
  required => 0,
  default => '/server-status?auto',
);

$options->arg(
  spec     => 'username|U=s',
  help     => 'username for basic auth',
  required => 0,
);

$options->arg(
  spec     => 'password|P=s',
  help     => 'password for basic auth',
  required => 0,
);

$options->arg(
  spec     => 'ssl|s',
  help     => 'use https instead of http',
  required => 0,
);

$options->arg(
  spec     => 'no_validate|N',
  help     => 'do not validate the SSL certificate chain',
  required => 0,
);

$options->arg(
  spec     => 'warning|w=s',
  help     => 'warning threshold',
  required => 0,
);

$options->arg(
  spec     => 'critical|c=s',
  help     => 'critical threshold',
  required => 0,
);

$options->arg(
  spec     => 'unreachable|R',
  help     => 'CRITICAL if socket timed out or http code >= 500',
  required => 0,
);

$options->getopts();

# if socket timed out or http code >= 500
my $unreachable = 'UNKNOWN';
if (defined($options->unreachable)) {
  $unreachable = 'CRITICAL';
}

alarm $options->timeout;
# override default alarm handler
$SIG{ALRM} = sub {
  $plugin->die(
    sprintf("plugin timed out (timeout %ss)",
      $options->timeout), $unreachable);
};


my @warning = split(",", $options->warning);
my @critical = split(",", $options->critical);

$threshold_OpenSlots = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[0],
  critical => $critical[0],
);
$threshold_BusyWorkers = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[1],
  critical => $critical[1],
);
$threshold_IdleWorkers = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[2],
  critical => $critical[2],
);
$threshold_ReqPerSec = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[3],
  critical => $critical[3],
);
$threshold_BytesPerSec = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[4],
  critical => $critical[4],
);
$threshold_BytesPerReq = Monitoring::Plugin::Threshold->set_thresholds(
  warning  => $warning[5],
  critical => $critical[5],
);

## Username without password
$plugin->nagios_exit( UNKNOWN, 'If you specify an username, you have to set a password too!') if ( ($options->username  ne '') && ($options->password eq '') );

## Password without username
$plugin->nagios_exit( UNKNOWN, 'If you specify a password, you have to set an username too!') if ( ($options->username  eq '') && ($options->password ne '') );

## Set account
if ( ($options->username ne '') && ($options->password ne '') ) {
  $account = $options->username.':'.$options->password.'@';
}

my $ua = LWP::UserAgent->new( protocols_allowed => ['http','https'], timeout => 15);
$ua->ssl_opts ( SSL_cipher_list => '' );

if (defined($options->no_validate)) {
  $ua->ssl_opts ( verify_hostname => 0 );
  $ua->ssl_opts ( SSL_verify_mode => 0 );
}

if (defined($options->ssl)) {
  $proto = 'https://';
} else {
  $proto = "http://";
}

if (defined($options->port)) {
  $request = HTTP::Request->new(GET => $proto.$account.$options->hostname.':'.$options->port.$options->uri.'?auto');
} else {
  $request = HTTP::Request->new(GET => $proto.$account.$options->hostname.$options->uri.'?auto');
}

$response = $ua->request($request);

if ($response->is_success) {

  unless ($response->content =~ /(?s).*ReqPerSec:\s([0-9\.]+).*BytesPerSec:\s([0-9\.]+).*BusyWorkers:\s([0-9]+).*IdleWorkers:\s([0-9]+).*Scoreboard:\s(\S*)/) {
    $plugin->plugin_exit( UNKNOWN, "No status information found at ".$response->base );
  }

  $ReqPerSec = $1;
  $BytesPerSec = $2;
  $BusyWorkers = $3;
  $IdleWorkers = $4;
  $OpenSlots   = ($5 =~ tr/[._]//);

  $response->content =~ /(?s).*BytesPerReq:\s([0-9\.]+)/;
  $BytesPerReq = $1;
  unless ($BytesPerReq) {
    $BytesPerReq = 0;
  }

  $output = 'OpenSlots:'.$OpenSlots.' BusyWorkers:'.$BusyWorkers.' IdleWorkers:'.$IdleWorkers.
    ' ReqPerSec:'.$ReqPerSec.' BytesPerSec:'.$BytesPerSec.' BytesPerReq:'.$BytesPerReq;

  $plugin->add_perfdata(
    label => 'OpenSlots',
    value => $OpenSlots,
    uom   => q{},
    threshold => $threshold_OpenSlots,
  );

  $plugin->add_perfdata(
    label => 'BusyWorkers',
    value => $BusyWorkers,
    uom   => q{},
    threshold => $threshold_BusyWorkers,
  );

  $plugin->add_perfdata(
    label => 'IdleWorkers',
    value => $IdleWorkers,
    uom   => q{},
    threshold => $threshold_IdleWorkers,
  );

  $plugin->add_perfdata(
    label => 'ReqPerSec',
    value => $ReqPerSec,
    uom   => q{},
    threshold => $threshold_ReqPerSec,
  );

  $plugin->add_perfdata(
    label => 'BytesPerSec',
    value => $BytesPerSec,
    uom   => B,
    threshold => $threshold_BytesPerSec,
  );

  $plugin->add_perfdata(
    label => 'BytesPerReq',
    value => $BytesPerReq,
    uom   => B,
    threshold => $threshold_BytesPerReq,
  );


  my @thresholds = (
    $threshold_OpenSlots->get_status($OpenSlots),
    $threshold_BusyWorkers->get_status($BusyWorkers),
    $threshold_IdleWorkers->get_status($IdleWorkers),
    $threshold_ReqPerSec->get_status($ReqPerSec),
    $threshold_BytesPerSec->get_status($BytesPerSec),
    $threshold_BytesPerReq->get_status($BytesPerReq)
  );

  my $status = 0;
  foreach(@thresholds) {
    if ($_ > $status) {
      $status = $_;
    }
  }

  $plugin->nagios_exit( $status, $output );

} elsif ( $response->code >= HTTP_INTERNAL_SERVER_ERROR ) {
  $plugin->plugin_exit( $unreachable, $response->status_line );
} else {
  $plugin->plugin_exit( UNKNOWN, $response->status_line );
}
